File: Days 21 - 30/Day 26/RemoveKthLastElementFromList.cpp
#include <iostream>

struct Node
{
	int value;
	Node* next;

	Node(int value, Node* next = nullptr)
		: value(value), next(next)
	{ }

	~Node() noexcept
	{
		if (next != nullptr)
		{
			delete next;			
		}
	}
};

void RemoveKthLastElementFromList(Node* head, int k) noexcept
{
	Node* front = head;
	Node* back = head;

	while (front != nullptr)
	{
		front = front->next;
		
		if (k < 0)
		{
			back = back->next;
		}

		k--;
	}

	if (k == 0)
	{
		head = head->next;
	}
	else
	{
		if (back->next != nullptr)
		{
			back->next = back->next->next;
		}
	}
}

void PrintLinkedList(Node* head) noexcept
{
	while (head != nullptr)
	{
		std::cout << head->value << " ";

		head = head->next;
	}

	std::cout << std::endl;
}

int main(int argc, const char* argv[])
{
	Node* n1 = new Node(4);
	Node* n2 = new Node(8, n1);
	Node* n3 = new Node(10, n2);
	Node* n4 = new Node(3, n3);
	Node* n5 = new Node(9, n4);
	Node* n6 = new Node(7, n5);
	Node* n7 = new Node(1, n6);
	Node* n8 = new Node(2, n7);

	PrintLinkedList(n8);

	RemoveKthLastElementFromList(n8, 3);

	PrintLinkedList(n8);
	
	std::cin.get();

	return 0;
}

File: Days 021 - 030/Day 26/RemoveKthLastElementFromList.cs
using System;

namespace DailyCodingProblem
{
	internal class Node
	{
		public int Value { get; private set; }
		public Node Next { get; set; }

		public Node(int value, Node next = null)
		{
			this.Value = value;
			this.Next = next;
		}
	}

	internal class Day26
	{
		private static int Main(string[] args)
		{
			Node n1 = new Node(4);
			Node n2 = new Node(8, n1);
			Node n3 = new Node(10, n2);
			Node n4 = new Node(3, n3);
			Node n5 = new Node(9, n4);
			Node n6 = new Node(7, n5);
			Node n7 = new Node(1, n6);
			Node n8 = new Node(2, n7);

			PrintLinkedList(n8);

			RemoveKthLastElementFromList(n8, 3);
			PrintLinkedList(n8);

			Console.ReadLine();

			return 0;
		}

		private static void RemoveKthLastElementFromList(Node head, int k)
		{
			Node front = head;
			Node back = head;

			while (front != null)
			{
				front = front.Next;

				if (k < 0)
				{
					back = back.Next;
				}

				k--;
			}

			if (k == 0)
			{
				head = head.Next;
			}
			else if (back.Next != null)
			{
				back.Next = back.Next.Next;
			}
		}

		private static void PrintLinkedList(Node head)
		{
			while (head != null)
			{
				Console.Write($"{head.Value} ");
				head = head.Next;
			}

			Console.WriteLine();
		}
	}
}